mypy/checker.py

```diff
@@ -4143,7 +4143,26 @@ def has_no_custom_eq_checks(t: Type) -> bool:
 
             return reduce_conditional_maps(partial_type_maps)
         elif isinstance(node, AssignmentExpr):
-            return self.find_isinstance_check_helper(node.target)
+            if_map = {}
+            else_map = {}
+
+            if_assignment_map, else_assignment_map = self.find_isinstance_check_helper(node.target)
+
+            if if_assignment_map is not None:
+                if_map.update(if_assignment_map)
+
+            if else_assignment_map is not None:
+                else_map.update(else_assignment_map)
+
+            if_condition_map, else_condition_map = self.find_isinstance_check_helper(node.value)
+
+            if if_condition_map is not None:
+                if_map.update(if_condition_map)
+
+            if else_condition_map is not None:
+                else_map.update(else_condition_map)
+
+            return if_map, else_map
```

Comment:
None values are special, but now we always return a non-`None` value. Should we sometimes still return a `None` value? The comment for `TypeMap` explains the distinction.

Reply:
Yes, we should return `None` value in case when `condition_map` is `None`. Basically there are three possibilities here:
- `assignment_map is None` and `condition_map is None`
```python
def always_false() -> Literal[False]: ...

if x := always_false():
    reveal_type(x)  # E: Statement is unreachable
```
- `assignment_map is not None` and `condition_map is None`

```python
if (x := 0):
    reveal_type(x)  # E: Statement is unreachable
```
- `assignment_map is not None` and `condition_map is not None`
```python
if (is_str := maybe_str is not None):
    reveal_type(is_str)  # N: Revealed type is 'builtins.bool'
    reveal_type(maybe_str)  # N: Revealed type is 'builtins.str'
```

I couldn't find any use case where `assignment_map is None` and `condition_map is not None`. I guess that there isn't actually such scenarios where this condition would be true. I added test cases for all above cases and corrected the return statement here.

Also, I modified slightly the behavior of `find_isinstance_check_helper` method to have a default mechanism for creating type maps (which was basically cut and pasted from the `RefExpr` case). This was actually needed to make use cases in `testWalrusAssignmentAndConditionScopeForFunction` test case work properly.